Sources/SourceKitLSP/DocumentTokens.swift

```diff
@@ -11,31 +11,92 @@
 //===----------------------------------------------------------------------===//
 
 import LanguageServerProtocol
+import SwiftSyntax
 
 /// Syntax highlighting tokens for a particular document.
 public struct DocumentTokens {
-  /// Lexical tokens, e.g. keywords, raw identifiers, ...
-  public var lexical: [SyntaxHighlightingToken] = []
+  /// The syntax tree representing the entire document.
+  public var syntaxTree: SourceFileSyntax?
   /// Semantic tokens, e.g. variable references, type references, ...
   public var semantic: [SyntaxHighlightingToken] = []
+}
 
-  private var merged: [SyntaxHighlightingToken] {
-    lexical.mergingTokens(with: semantic)
-  }
-  public var mergedAndSorted: [SyntaxHighlightingToken] {
-    merged.sorted { $0.start < $1.start }
+extension DocumentSnapshot {
+  private func highlightingToken(from syntax: SyntaxClassifiedRange) -> [SyntaxHighlightingToken] {
+    guard let (kind, modifiers) = syntax.kind.parseKindAndModifiers else {
+      return []
+    }
+
+    guard
+      let start: Position = self.positionOf(utf8Offset: syntax.offset),
+      let end: Position = self.positionOf(utf8Offset: syntax.endOffset)
+    else {
+      return []
+    }
+
+    let multiLineRange = start..<end
+    let ranges = multiLineRange.splitToSingleLineRanges(in: self)
+
+    return ranges.map {
+      SyntaxHighlightingToken(
+        range: $0,
+        kind: kind,
+        modifiers: modifiers
+      )
+    }
   }
 
-  /// Modifies the syntax highlighting tokens of each kind
-  /// (lexical and semantic) according to `action`.
-  public mutating func withMutableTokensOfEachKind(_ action: (inout [SyntaxHighlightingToken]) -> Void) {
-    action(&lexical)
-    action(&semantic)
+  /// Computes an array of syntax highlighting tokens from the syntax tree that
+  /// have been merged with any semantic tokens from SourceKit. If the provided
+  /// range is non-empty, this function restricts its output to only those
+  /// tokens whose ranges overlap it. If no range is provided, tokens for the
+  /// entire document are returned.
+  ///
+  /// - Parameter range: The range of tokens to restrict this function to, if any.
+  /// - Returns: An array of syntax highlighting tokens.
+  public func mergedAndSortedTokens(in range: Range<Position>? = nil) -> [SyntaxHighlightingToken] {
+    guard let tree = self.tokens.syntaxTree else {
+      return self.tokens.semantic
+    }
+    let range = range.flatMap({ self.utf8OffsetRange(of: $0) }).map({ ByteSourceRange(offset: $0.startIndex, length: $0.count) })
+             ?? ByteSourceRange(offset: 0, length: tree.byteSize)
```

Comment:
Do you think it would be worth putting this in an extension on `Range<Position>`. It seems to me like we’ll be needing it more often in the future.

Something like

```swift
  func byteSourceRange(in snapshot: DocumentSnapshot): ByteSourceRange {
    return = self.flatMap({ snapshot.utf8OffsetRange(of: $0) }).map({ ByteSourceRange(offset: $0.startIndex, length: $0.count) })
             ?? ByteSourceRange(offset: 0, length: snapshot.tree.byteSize)
  }
```